### PR TITLE
fix(cake): quoted references

### LINKED ISSUE #9691: Cake referenced are ignored if they are in quotes
**What Renovate type, platform and version are you using?**

GitHub

**Describe the bug**

The `Cake` manager that extracts references ignores references that are enclosed in quotes.

the two following lines are equivalent, as far as Cake is concerned:
```
#addin "nuget:?package=Foo.Bar&version=1"
#addin nuget:?package=Foo.Bar&version=1
``` 

However, the first line where the reference-part is set in quotes is ignored by the manager.

A minimal sample is available at https://github.com/nils-a-demos/renovate-cake-quotes-1
The minimal sample referenced two separate packages, `cake.7zip` and `cake.asciidoctorj` (the latter is set in quotes).

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
DEBUG: Found cake package files
DEBUG: Found 1 package file(s)
INFO: Dependency extraction complete
{
  "baseBranch": "main",
  "stats": {
    "managers": {
      "cake": {
        "fileCount": 1,
        "depCount": 1
      }
    },
    "total": {
      "fileCount": 1,
      "depCount": 1
    }
  }
}
DEBUG: Package releases lookups complete
{
  "baseBranch": "main"
}
DEBUG: branchifyUpgrades
DEBUG: 2 flattened updates found: Cake.7zip, Cake.7zip
DEBUG: Returning 2 branch(es)
DEBUG: Fetching changelog: https://github.com/cake-contrib/Cake.7zip (0.1.0 -> 0.7.1)
DEBUG: Fetching changelog: https://github.com/cake-contrib/Cake.7zip (0.1.0 -> 1.0.3)
DEBUG: config.repoIsOnboarded=true
DEBUG: packageFiles with updates
{
  "config": {
    "cake": [
      {
        "packageFile": "build.cake",
        "deps": [
          {
            "datasource": "nuget",
            "depName": "Cake.7zip",
            "currentValue": "0.1.0",
            "depIndex": 0,
            "updates": [
              {
                "newVersion": "0.7.1",
                "newValue": "0.7.1",
                "bucket": "non-major",
                "newMajor": 0,
                "newMinor": 7,
                "updateType": "minor",
                "releaseTimestamp": "2020-10-07T21:27:57.923Z",
                "branchName": "renovate/cake.7zip-0.x"
              },
              {
                "newVersion": "1.0.3",
                "newValue": "1.0.3",
                "bucket": "major",
                "newMajor": 1,
                "newMinor": 0,
                "updateType": "major",
                "releaseTimestamp": "2021-04-10T21:38:10.890Z",
                "branchName": "renovate/cake.7zip-1.x"
              }
            ],
            "warnings": [],
            "sourceUrl": "https://github.com/cake-contrib/Cake.7zip",
            "homepage": "https://cake-contrib.github.io/Cake.7zip",
            "currentVersion": "0.1.0",
            "isSingleVersion": true,
            "fixedVersion": "0.1.0"
          }
        ]
      }
    ]
  }
}
DEBUG: processRepo()
DEBUG: Processing 2 branches: renovate/cake.7zip-0.x, renovate/cake.7zip-1.x
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [X] I have provided a minimal reproduction repository
- [ ] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

I'll add a PR to update lexer and parser in the `Cake` manager.

### PR DESCRIPTION
**Pull request by nils-a**
## Changes:

I modified parser and lexer of the Cake manager to
recognize references that are fully set in quotes.

## Context:

Closes #9691 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
LGTM, @zharinov can you check the lexer changes?

File: lib/manager/cake/index.ts [hunk 1]
```diff
@@ -17,15 +17,23 @@ const lexerStates = {
     dependency: {
       match: /^#(?:addin|tool|module)\s+(?:nuget|dotnet):.*$/,
     },
+    dependencyQuoted: {
+      match: /^#(?:addin|tool|module)\s+"(?:nuget|dotnet):[^"]+"\s*$/,
```
Comment (viceice) [inline lib/manager/cake/index.ts:21]: We should add a moo [transform}(https://github.com/no-context/moo#value-vs-text) function, so we don't need to care about the quotes later

```suggestion
      match: /^#(?:addin|tool|module)\s+"(?:nuget|dotnet):[^"]+"\s*$/,
      value: s => s.slice(1, -1),
```

File: lib/manager/cake/index.ts [hunk 2]
```diff
@@ -17,15 +17,23 @@ const lexerStates = {
...
+    },
     unknown: { match: /[^]/, lineBreaks: true },
   },
 };
 
 function parseDependencyLine(line: string): PackageDependency | null {
   try {
-    let url = line.replace(/^[^:]*:/, '');
+    let url = line.replace(/^[^:]*:/, '').trimEnd();
```
Comment (viceice) [inline lib/manager/cake/index.ts:29]:
```suggestion
    let url = line.replace(/^[^:]*:/, '');
```

File: lib/manager/cake/index.ts [hunk 3]
```diff
@@ -17,15 +17,23 @@ const lexerStates = {
...
     const isEmptyHost = url.startsWith('?');
     url = isEmptyHost ? `http://localhost/${url}` : url;
+
+    if (url.endsWith('"')) {
+      url = url.slice(0, -1);
+    }
+
```
Comment (viceice) [inline lib/manager/cake/index.ts:36]:
```suggestion
```

**Review by viceice — CHANGES_REQUESTED** [review]
I think this should simplify it.

**Comment by viceice** [comment]
Even if unrelated, please move const lexer to module scope for perfomance and memory reasons.

https://github.com/renovatebot/renovate/blob/ed0605ebc3af49264b2a51dc197dfc4d9314d793/lib/manager/cake/index.ts#L52

@rarkins I think we (re)create too many moo lexers, which can be another memory and performance issue. So @zharinov can you open a pr to move those to module scope, so the lexers are created once?

**Comment by rarkins** [comment]
Does anything need to be changed for this PR to be accepted though?

**Comment by viceice** [comment]
> Does anything need to be changed for this PR to be accepted though?

See my second review 😉

File: lib/manager/cake/__fixtures__/build.cake [hunk 4]
```diff
@@ -1,5 +1,6 @@
 foo
 #addin nuget:?package=Foo.Foo&version=1.1.1
+#addin "nuget:?package=Bim.Bim&version=6.6.6"
```
Comment (zharinov) [inline lib/manager/cake/__fixtures__/build.cake:3]: What about other quote styles? E.g. `"""foobar"""`, `'''foobar'''` or ``  `foobar` ``

**Comment by zharinov** [comment]
@viceice I'll move all lexers to global scope in separate PR

**Review by zharinov — DISMISSED** [review]
LGTM

**Comment by viceice** [comment]
@zharinov What do you think about my suggestions to use a moo transform instead of triming the quotes later?

File: lib/manager/cake/index.ts (hunk 1 above)
Comment (zharinov) [reply lib/manager/cake/index.ts:21]:
```ts
s.replace(/"\s*$/, '')
```

Comment (zharinov) [reply lib/manager/cake/index.ts:21]: We also can remove `#prefix nuget:` part, though we may want to extract dep type some time later

File: lib/manager/cake/__fixtures__/build.cake (hunk 4 above)
Comment (nils-a) [reply lib/manager/cake/__fixtures__/build.cake:3]: The are not documented and I've never ever seen anything other than `"` out in the wilds. Will run some quick checks to be sure.

File: lib/manager/cake/index.ts (hunk 1 above)
Comment (nils-a) [reply lib/manager/cake/index.ts:21]: love it. Seems I need to read up on moo 😄 
But simply slicing leads to errors for trailing spaced after the last quote, right?

Comment (viceice) [reply lib/manager/cake/index.ts:21]: Yes, so needs trim before slice.

**Comment by nils-a** [comment]
@viceice I implemented the suggested change from `if` to a moo transform.

**Review by zharinov — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
@rarkins Backport to v24?

**Pull request merged**